src/ConstantSum/ConstantSum.sol

```diff
@@ -58,21 +65,106 @@ contract ConstantSum is PairStrategy {
         )
     {
         ConstantSumParams memory params;
-        (reserves, totalLiquidity, params) =
-            abi.decode(data, (uint256[], uint256, ConstantSumParams));
+
+        (reserves, params) = abi.decode(data, (uint256[], ConstantSumParams));
+        totalLiquidity =
+            computeDeltaLiquidity(reserves[0], reserves[1], params.price);
+
+        if (pool.reserves.length != 2 || reserves.length != 2) {
+            revert InvalidReservesLength();
+        }
 
         internalParams[poolId].price = params.price;
         internalParams[poolId].swapFee = params.swapFee;
+        internalParams[poolId].controller = params.controller;
 
         // Get the trading function and check this is valid
         invariant =
             tradingFunction(reserves, totalLiquidity, abi.encode(params));
 
-        valid = invariant >= 0;
+        valid = invariant >= 0 && invariant <= EPSILON;
 
         return (valid, invariant, reserves, totalLiquidity);
     }
 
+    function validateAllocate(
+        address,
+        uint256 poolId,
+        Pool memory pool,
+        bytes calldata data
+    )
+        external
+        view
+        override
+        returns (
+            bool valid,
+            int256 invariant,
+            uint256[] memory deltas,
+            uint256 deltaLiquidity
+        )
+    {
+        (uint256 deltaX, uint256 deltaY, uint256 minDeltaL) =
+            abi.decode(data, (uint256, uint256, uint256));
+
+        deltaLiquidity =
+            computeDeltaLiquidity(deltaX, deltaY, internalParams[poolId].price);
```

Comment:
It's worth noting that `computeDeltaLiquidity()` does `price.mulWadUp(deltaX)`, which rounds up. 

As such, the rounding direction here actually favors the user as it rounds away from the slippage parameter (ie. the user declares a minimum amount of `deltaLiquidity` acceptable, but the rounding makes `deltaLiquidity` larger instead of smaller).

Contrast this with `validateAllocate()` in other strategies, such as `GeometricMean`, where the user provides a maximum acceptable amount of `tokenX` and `tokenY` that needs to be transferred into the pool, and the amount of tokens calculated rounds up towards the maximum amount specified:

```solidity
    function _computeAllocateDeltasGivenDeltaL(
        uint256 deltaLiquidity,
        Pool memory pool,
        bytes memory
    ) internal pure override returns (uint256[] memory deltas) {
        deltas = new uint256[](2);
        deltas[0] = computeDeltaGivenDeltaLRoundUp(
            pool.reserves[0], deltaLiquidity, pool.totalLiquidity
        );

        deltas[1] = computeDeltaGivenDeltaLRoundUp(
            pool.reserves[1], deltaLiquidity, pool.totalLiquidity
        );
    }
```

The effect of this is that you mint 1 wei extra of `deltaLiquidity` for the amount of `tokenX` transferred in.

The impact is that if:
- The current invariant is near `0`.
- `validateAllocate()` is called with `price * deltaX` not exactly divisible by `1e18`.

`deltaLiquidity` will be 1 wei larger than it should be, making it possible for the invariant to go negative and fail the invariant check. For example:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.18;

import "forge-std/Test.sol";
import "src/FixedPointMathLib.sol";

contract AllocateTest is Test {
    using FixedPointMathLib for uint256;

    function testAllocate() public {
        // Pool state
        uint256 rX = 70e18; // DAI
        uint256 rY = 1e6;   // WBTC
        uint256 price = uint256(1e6) * 1e18 / 70e18;

        uint256 tL = price.mulWadUp(rX) + rY;
        _invariant(rX, rY, tL, price);

        // Allocate 1e18 + 1 DAI
        uint256 deltaX = 1e18 + 1;
        uint256 deltaY = 0;
        
        uint256 dL = computeDeltaLiquidity(deltaX, deltaY, price);
        _invariant(rX + deltaX, rY + deltaY, tL + dL,price);
    }

    function _invariant(
        uint256 rX,
        uint256 rY,
        uint256 tL,
        uint256 price
    ) internal returns (int256) {
        int256 invariant = int256(price.mulWadUp(rX.divWadUp(tL)) + rY.divWadUp(tL)) - int256(1e18);
        console2.log("invariant:", invariant);
    }

    function computeDeltaLiquidity(
        uint256 deltaX,
        uint256 deltaY,
        uint256 price
    ) internal returns (uint256) {
        return price.mulWadUp(deltaX) + deltaY;
    }
}
```

Output:
```
Logs:
  invariant: 1
  invariant: -496466153916
```

Recommende fix would be to calculate `deltaLiquidity` as `price.mulWadDown(deltaX) + deltaY` for **only** `validateAllocate()`.

Reply:
I've introduced another function to compute the liquidity and round up or down accordingly, see: https://github.com/primitivefinance/DFMM/pull/119/commits/1e51776c4a66d2d627adfac587cf25cf7972a5bd.